### PR TITLE
feat(hotkey): dynamic change some parameters in hotkey detection

### PR DESCRIPTION
**Pull request by Smityz**

### CODE REVIEW

**Review by Shuo-Jia — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Review by neverchanje — APPROVED** [review]

File: src/server/hotspot_partition_calculator.cpp [hunk 1]
```diff
@@ -41,17 +41,20 @@ DSN_DEFINE_bool("pegasus.collector",
...
                 false,
                 "auto detect hot key in the hot paritition");
+DSN_TAG_VARIABLE(enable_detect_hotkey, FT_MUTABLE);
 
 DSN_DEFINE_int32("pegasus.collector",
                  hot_partition_threshold,
                  3,
                  "threshold of hotspot partition value, if app.stat.hotspots >= "
                  "FLAGS_hotpartition_threshold, this partition is a hot partition");
+DSN_TAG_VARIABLE(hot_partition_threshold, FT_MUTABLE);
```
Comment (acelyc111) [inline src/server/hotspot_partition_calculator.cpp:51]:
```
    for (int index = 0; index < _hot_points.size(); index++) {
        if (_hot_points[index][data_type].get()->get_value() >= FLAGS_hot_partition_threshold) {
            if (++_hotpartition_counter[index][data_type] >= FLAGS_occurrence_threshold) {
```
Better to move hot_partition_threshold and occurrence_threshold out of loop.

**Comment by hycdong** [comment]
How about add validator for the configuration which is necessary to validate?

**Review by acelyc111 — DISMISSED** [review]

**Review by levy5307 — APPROVED** [review]

File: src/server/hotkey_collector.cpp [hunk 2]
```diff
@@ -33,12 +33,18 @@ DSN_DEFINE_uint32(
     hot_bucket_variance_threshold,
     7,
     "the variance threshold to detect hot bucket during coarse analysis of hotkey detection");
+DSN_TAG_VARIABLE(hot_bucket_variance_threshold, FT_MUTABLE);
+DSN_DEFINE_validator(hot_bucket_variance_threshold,
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:37]: uint32 must be >= 0, I think it's useless to add a validator here.

File: src/server/hotspot_partition_calculator.cpp [hunk 3]
```diff
@@ -41,17 +41,24 @@ DSN_DEFINE_bool("pegasus.collector",
                 enable_detect_hotkey,
                 false,
                 "auto detect hot key in the hot paritition");
+DSN_TAG_VARIABLE(enable_detect_hotkey, FT_MUTABLE);
 
 DSN_DEFINE_int32("pegasus.collector",
```
Comment (levy5307) [inline src/server/hotspot_partition_calculator.cpp:46]: DSN_DEFINE_int32 -> DSN_DEFINE_uint32

File: src/server/hotspot_partition_calculator.cpp [hunk 4]
```diff
@@ -41,17 +41,24 @@ DSN_DEFINE_bool("pegasus.collector",
...
                  hot_partition_threshold,
                  3,
                  "threshold of hotspot partition value, if app.stat.hotspots >= "
                  "FLAGS_hotpartition_threshold, this partition is a hot partition");
+DSN_TAG_VARIABLE(hot_partition_threshold, FT_MUTABLE);
+DSN_DEFINE_validator(hot_partition_threshold,
+                     [](int32_t threshold) -> bool { return threshold >= 0; });
 
 DSN_DEFINE_int32("pegasus.collector",
```
Comment (levy5307) [inline src/server/hotspot_partition_calculator.cpp:55]: DSN_DEFINE_int32 -> DSN_DEFINE_uint32

**Review by levy5307 — APPROVED** [review]

**Review by hycdong — APPROVED** [review]

**Pull request merged**